### PR TITLE
CI Improvements - Part 1

### PR DESCRIPTION
**Pull request by SergioGasquez**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/CONTRIBUTING.md) and followed its instructions.

#### Description
- HIL workflow: Fix the condition to erase flash when tests fail. [Source](https://docs.github.com/en/actions/learn-github-actions/expressions#example-of-failure-with-conditions)
- Remove `path-ignore` to avoid CI not being triggered.
- Build documentation

#### Testing
- The HIL workflow cant be tested easily, we need to monitor the workflow and verify that upon a test failure, the erase-flash step is executed.
- Documentation was built in my fork https://github.com/SergioGasquez/esp-hal/actions/runs/9972410348/attempts/1. I ran the workflow twice, both took ~10min45. Currently, CI takes between 10 to 14 mins, so I guess we didn't increased it too much. 
  - Happy to move this to a nightly CI workflow, but it looks like it doesnt add much time.

### CODE REVIEW

**Review by bjoernQ — APPROVED** [review]
LGTM - I don't really understand the fix for the flash-erase but that's most probably a lack of knowledge about GitHub workflows on my side

**Comment by SergioGasquez** [comment]
> LGTM - I don't really understand the fix for the flash-erase but that's most probably a lack of knowledge about GitHub workflows on my side

I don't fully understand it either, but it [wasn't working before](https://github.com/esp-rs/esp-hal/actions/runs/9957400001/job/27509536912) and the GH docs has an example [using that same condition](https://docs.github.com/en/actions/learn-github-actions/expressions#example-of-failure-with-conditions)

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -113,6 +113,8 @@ jobs:
           components: rust-src
       - name: Check doc-tests
         run: cargo xtask run-doc-test esp-hal ${{ matrix.device.soc }}
+      - name: Check documentation
+        run:  cargo xtask build-documentation --packages esp-hal --chips ${{ matrix.device.soc }}
```
Comment (MabezDev) [inline .github/workflows/ci.yml:117]: When we build the docs, are we checking that we have no warnings? Not checking warnings makes this step not that useful I think, as it's very unlikely that docs are left in an unbuildable state.

Comment (SergioGasquez) [reply .github/workflows/ci.yml]: > When we build the docs, are we checking that we have no warnings? 

Now we are! Good sugestion

**Review by MabezDev — APPROVED** [review]
Thanks, this is a nice improvement!

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -70,13 +64,19 @@ jobs:
     steps:
       - uses: actions/checkout@v4
 
-      # Install the Rust toolchain for RISC-V devices:
+      # Install the Rust stable and nightly toolchains for RISC-V devices:
```
Comment (jessebraham) [inline .github/workflows/ci.yml:67]: This appears to just install both toolchains, and not build with both of them? Or am I missing something here?

**Comment by AnthonyGrondin** [comment]
For the flash-erase fix, it seems to be because Github's implementation require `failure()` to be present in the if statement, else it automatically uses `success()`

> [failure with conditions 🔗](https://docs.github.com/en/actions/learn-github-actions/expressions#failure-with-conditions)
> You can include extra conditions for a step to run after a failure, but you must still include `failure()` to override the default status check of `success()` that is automatically applied to `if` conditions that don't contain a status check function.

**Pull request merged**